File: src/main/java/sx/jms/SimpleListener.java
package sx.jms;

import javax.jms.*;
import java.util.List;

/**
 * Simple single threaded jms listener
 * Created by masc on 17.04.15.
 */
public abstract class SimpleListener extends Listener {
    Connection mConnection;
    Session mSession;

    public SimpleListener(ConnectionFactory connectionFactory) {
        super(connectionFactory);
    }

    protected Connection createConnection() throws JMSException {
        return this.getConnectionFactory().createConnection();
    }

    protected abstract Session createSession(Connection connection) throws JMSException;

    protected abstract List<Destination> createDestinations(Session session) throws JMSException;

    protected Connection getConnection() {
        return mConnection;
    }

    protected Session getSession() {
        return mSession;
    }

    @Override
    public void start() throws JMSException {
        this.stop();

        mConnection = this.createConnection();
        mConnection.start();

        mSession = this.createSession(mConnection);

        List<Destination> dests = this.createDestinations(mSession);
        for (Destination d : dests) {
            MessageConsumer mc = mSession.createConsumer(d);

            // Wrap message callback, adding jms transaction support
            mc.setMessageListener( (message) -> {
                boolean transacted = false;
                try {
                    transacted = mSession.getTransacted();
                    SimpleListener.this.onMessage(message);
                    if (mSession.getTransacted())
                        mSession.commit();
                } catch(Exception e) {
                    // TODO: verify if exception is routed to onException handler when not caught here
                    this.getLog().error(e.getMessage(), e);

                    if (transacted) {
                        try {
                            mSession.rollback();
                        } catch (JMSException e1) {
                            this.getLog().error(e.getMessage(), e);
                        }
                    }
                }
            });
        }
    }

    @Override
    public void stop() throws JMSException {
        if (this.mConnection != null)
            mConnection.close();
    }
}

File: libs/sx-mq/src/main/java/sx/jms/listeners/SimpleListener.kt
package sx.jms.listeners

import sx.LazyInstance
import sx.jms.Channel
import sx.jms.Listener
import javax.jms.*

/**
 * Simple single threaded jms listener
 * Created by masc on 17.04.15.
 * @param connectionFactory Connection factory
 * @param converter Message converter
 */
open class SimpleListener(channel: () -> Channel)
    :
        Listener(channel) {

    private val connection = LazyInstance<Connection>({
        this.channel.connectionFactory!!.createConnection()
    })

    private val session = LazyInstance<Session>({
        this.connection.get().createSession(true, Session.AUTO_ACKNOWLEDGE)
    })

    private val consumer = LazyInstance<MessageConsumer>({
        this.session.get().createConsumer(this.channel.destination)
    })

    /**
     * Start listener
     * @throws JMSException
     */
    @Throws(JMSException::class)
    @Synchronized final override fun start() {
        this.stop()

        // Wrap message callback, adding jms transaction support
        this.consumer.get().messageListener = object : MessageListener {
            override fun onMessage(message: Message) {
                val session = this@SimpleListener.session.get()
                try {
                    this@SimpleListener.onMessage(message, session)
                    if (session.transacted)
                        session.commit()
                } catch (e: Exception) {
                    // TODO: verify if exception is routed to onException handler when not caught here
                    log.error(e.message, e)

                    if (session.transacted) {
                        try {
                            session.rollback()
                        } catch (e1: JMSException) {
                            log.error(e.message, e)
                        }
                    }
                }
            }
        }

        this.connection.get().start()
    }

    /**
     * Stop listener
     * @throws JMSException
     */
    @Throws(JMSException::class)
    @Synchronized final override fun stop() {
        this.consumer.ifSet {
            it.close()
        }

        this.session.ifSet {
            it.close()
        }

        this.connection.ifSet {
            it.close()
        }
    }
}
